File: sample_algorithms/Sorting/python/bbsort.py
import tracers.tracer as trc

# [ALGORITHM]
def bubble_sort(arr, tracer):
    n = len(arr)
    for i in range(n):
        tracer.add_state(arr.copy())  # Track state
        swapped = False
        for j in range(0, n - i - 1):
            # Show comparison (yellow)
            tracer.add_state(arr.copy(), comparing=[j, j + 1])
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                # Show swap result (green)
                tracer.add_state(arr.copy(), swapped=[j, j + 1])
                swapped = True
        if not swapped:
            break
    return arr, tracer

# [TEST]
if __name__ == "__main__":
    original_arr = [92, 14, 461, 1122, 235, 9, 127]
    sorted_arr, tracer = bubble_sort(original_arr.copy(), trc.Tracer(category='sorting'))

    print(f'Original array: {original_arr}')
    print(f'Sorted array: {sorted_arr}')
    
    # Output tracer data for visualization
    tracer.finalize()

File: sample_algorithms/Sorting/javascript/bbsort.js
const Tracer = require('./tracers/tracer');

// [ALGORITHM]
function bubbleSort(arr, tracer) {
    const n = arr.length;
    
    for (let i = 0; i < n; i++) {
        tracer.addState([...arr]);  // Track state
        let swapped = false;
        
        for (let j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                [arr[j], arr[j + 1]] = [arr[j + 1], arr[j]];
                tracer.addState([...arr]);  // Track state
                swapped = true;
            }
        }
        
        if (!swapped) {
            break;
        }
    }
    
    return [arr, tracer];
}

// [TEST]
if (require.main === module) {
    const originalArr = [64, 34, 25, 12, 22, 11, 90];
    const tracer = new Tracer('sorting'); // Tracer instance
    const [sortedArr] = bubbleSort([...originalArr], tracer);
    
    console.log('Original array:', originalArr);
    console.log('Sorted array:', sortedArr);
    
    // Output tracer data for visualization
    tracer.finalize();
}